### DISCUSSION TITLE
pairing to bluetooth OBD-II reader

### DISCUSSION THREAD

**Discussion by mhixson** [micropython/micropython#9490] [Hardware & Peripherals]
Does anyone here have experience getting an ESP32 based board to pair with an OBD-II reader over Bluetooth?

Specifically, I'd like to connect a [Sparkfun Thing Plus C](https://www.sparkfun.com/products/18018) to a [Scantool MX+](https://www.obdlink.com/products/obdlink-mxp/). I've seen several projects making use of a cable to do this, but haven't found anyone using Bluetooth. Anyone here have experience with this?
 Thanks.

**Comment by jimmo** [comment]
This thread from the old forum might be useful -- https://forum.micropython.org/viewtopic.php?f=2&t=10812

Note the issue at the end about messages arriving too quickly, this may now be solved in aioble with the "capture" option on characteristics.

**Reply by mhixson** [reply]
Thanks, for the reply and the interesting thread. However, that's far lower level than where I currently am.

At this point I'd like to just see a working Bluetooth example that runs on micropython on my Thing Plus that shows me that it can communicate with the OBD-II reader.

On my Raspberry Pi I have connected to the OBD-II reader with this:

sudo rfcomm bind 0 00:04:3E:5A:9E:E9

Which creates a /dev/rfcomm0 serial device.

Using micropython how can I verify that I can communicate and/or pair with that reader?

I'm using [python-obd](https://python-obd.readthedocs.io/en/latest/) on my Pi so I also need to learn how to install python modules that aren't baked into the main firmware. Which leads me to [this documentation](https://docs.micropython.org/en/latest/reference/packages.html#installing-packages-with-mpremote).

MicroPython v1.19.1 on 2022-06-18; ESP32 module with ESP32
Type "help()" for more information.
>>> import mip
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
ImportError: no module named 'mip'

So I try using mpremote:

$ mpremote mip install obd
mpremote: 'mip' is not a command

And I have also installed mip:
$ pip install mip
Requirement already satisfied: mip in /Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/site-packages (1.14.1)
Requirement already satisfied: cffi==1.15.0 in /Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/site-packages (from mip) (1.15.0)
Requirement already satisfied: pycparser in /Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/site-packages (from cffi==1.15.0->mip) (2.21)

Any ideas on how to go about installing packages onto the ESP32?
 Thanks again.

**Reply by jimmo** [reply]
> Which leads me to [this documentation](https://docs.micropython.org/en/latest/reference/packages.html#installing-packages-with-mpremote).

This is the "latest" documentation which refers to the in-development version (i.e. the nightly builds). `mip` will be in the next release (v1.20)...soon. Most of the time the documentation is pretty close, but you happen to be looking at a page that was literally updated this week for the first time in a couple of years.

(The v1.19.1 page is here: https://docs.micropython.org/en/v1.19.1/reference/packages.html )

(I think we should probably come up with some way to make the docs point to the current release version by default... @dpgeorge what do you think? )

> Which creates a /dev/rfcomm0 serial device.

rfcomm suggests that this is Bluetooth Classic, not BLE though. We only support BLE.

Taking a quick look at the obdlink page, I think only the "OBDLINK® CX" is a BLE device.

> I'm using [python-obd](https://python-obd.readthedocs.io/en/latest/) on my Pi

Also, any random Python package (e.g. `obd` in this case) is unlikely to work directly on MicroPython. This is for a few different reasons -- either supported OS features (e.g. on Linux this is requiring pyserial to talk to the rfcomm device), or for language features.

This is actually a big part of why we've removed PyPI package support from the MicroPython package manager. Instead we now have our own dedicated package index of things known to work.

**Reply by mhixson** [reply]
> > Which leads me to [this documentation](https://docs.micropython.org/en/latest/reference/packages.html#installing-packages-with-mpremote).
> 
> This is the "latest" documentation which refers to the in-development version (i.e. the nightly builds). `mip` will be in the next release (v1.20)...soon. Most of the time the documentation is pretty close, but you happen to be looking at a page that was literally updated this week for the first time in a couple of years.
> 
> (The v1.19.1 page is here: https://docs.micropython.org/en/v1.19.1/reference/packages.html )

Ah, that explains my confusion. The documentation I was looking at and my MicroPython header matched so I was expecting that to work. Anyhow, I found some other docs recommending upip. I was able to get obd installed using that, however I get this just on trying to import it:

MicroPython v1.19.1 on 2022-06-18; ESP32 module with ESP32
Type "help()" for more information.
import obd
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/lib/obd/__init__.py", line 41, in <module>
  File "/lib/obd/obd.py", line 36, in <module>
  File "/lib/obd/OBDResponse.py", line 38, in <module>
MemoryError: memory allocation failed, allocating 3136 bytes

> > Which creates a /dev/rfcomm0 serial device.
> 
> rfcomm suggests that this is Bluetooth Classic, not BLE though. We only support BLE.

Ok, that fact combined with the memory allocation problem above is starting to make me think that this ESP32 platform is not the tool for the job and I'd probably save myself a lot of pain and anguish if I went with a Raspberry Pi Zero 2 W instead.

Thanks for your help and explanation about all of that.
 -Matt

**Comment by zefoo** [comment]
I'd be curious if you ever got this working @mhixson ?

**Reply by mhixson** [reply]
I did not get it working with ESP32. At this time I gave up on that and moved to a Raspberry Pi 4. My life got a whole lot easier at that point.

If someone in the future does find a way to get the ESP32 or similar Arduino chips connected to a Bluetooh OBD-II reader I'd love to hear about it.
